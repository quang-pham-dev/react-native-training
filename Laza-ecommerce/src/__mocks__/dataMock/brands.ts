/* A data array. */

export const brand = {
  id: '0',
  name: 'Nike',
  logoUrl:
    'https://toppng.com/uploads/preview/ike-logo-png-hd-nike-logo-png-image-free-download-nike-logo-sv-11562865034fkhqkcoas6.png'
};

export const brands = [
  {
    id: '0',
    name: 'Nike',
    logoUrl:
      'https://toppng.com/uploads/preview/ike-logo-png-hd-nike-logo-png-image-free-download-nike-logo-sv-11562865034fkhqkcoas6.png'
  },
  {
    id: '1',
    name: 'Adidas',
    logoUrl:
      'https://www.pikpng.com/pngl/m/42-429866_adidas-logo-png-images-free-download-rh-pngimg.png'
  },
  {
    id: '2',
    name: 'Fila',
    logoUrl:
      'https://toppng.com/uploads/preview/fila-logo-vector-download-free-11574216707jcmvfu58fy.png'
  },
  {
    id: '3',
    name: 'Puma',
    logoUrl:
      'https://banner2.cleanpng.com/20180613/gsj/kisspng-tracksuit-puma-logo-brand-clothing-puma-logo-5b20dc6b232426.487874081528880235144.jpg'
  },
  {
    id: '4',
    name: 'Gucci',
    logoUrl: 'https://esquilo.io/png/thumb/XBwczcMncYeff65-Vector-Gucci-Logo-PNG-Free-Download.png'
  }
];
