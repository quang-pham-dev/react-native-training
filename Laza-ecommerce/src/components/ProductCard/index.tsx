import React, { memo, useCallback } from 'react';
import { Image, ImageBackground, Pressable, Text, View } from 'react-native';

// LIBS
import isEqual from 'react-fast-compare';

// Components
import Title from 'components/Title';

// Types
import { IProductCardProps } from 'types/models/Products';

// Themes
import IMAGES from 'themes/Images';

// Styles
import styles from './styles';

const ProductCard = ({
  product,
  onPressProductCard,
  onPressLikeProduct,
  productCardStyles
}: IProductCardProps) => {
  // handle action press product card with id
  const handlePressCardProduct = useCallback(() => {
    onPressProductCard(product.id);
  }, [product.id, onPressProductCard]);

  // handle action press like product
  const handlePressLikeProduct = useCallback(() => {
    onPressLikeProduct(product);
  }, [product, onPressLikeProduct]);

  return (
    <Pressable
      testID='productCard'
      onPress={handlePressCardProduct}
      style={[styles.productCardContainer, productCardStyles]}
    >
      <View style={styles.imageWrapper}>
        <ImageBackground style={styles.image} source={{ uri: product?.source }} />
        <Pressable onPress={handlePressLikeProduct} style={styles.iconHeartWrapper}>
          <Image
            style={styles.iconHeart}
            source={product?.like ? IMAGES.iconHeartLiked : IMAGES.iconHeart}
          />
        </Pressable>
      </View>
      <Title titleStyles={styles.productTitle} titleName={product?.title} />
      <Title titleStyles={styles.productType} titleName={product?.type} />
      <Text style={styles.price}>{`$ ${product?.price}`}</Text>
    </Pressable>
  );
};

export default memo(ProductCard, isEqual);
