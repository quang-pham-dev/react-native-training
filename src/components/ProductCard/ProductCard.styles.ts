import styled from 'styled-components/native'
import {s} from 'react-native-size-matters/extend'

// Themes
import {Colors, Metrics} from '@themes'

export const ProductCardWrapperStyled = styled.Pressable`
  width: ${s(160)}px;
  heigh: ${s(260)}px;
  background-color: ${Colors.background};
  justify-content: center;
`
export const ImageWrapperStyled = styled.View`
  border-radius: ${Metrics.borderRadius.medium}px;
  margin-bottom: ${Metrics.margin.tiny}px;
  overflow: hidden;
`

export const IconHeartWrapperStyled = styled.Pressable`
  position: absolute;
  top: 14px;
  right: 15px;
  z-index: 1;
`
