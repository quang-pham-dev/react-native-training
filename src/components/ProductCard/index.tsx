import React, {memo, useCallback} from 'react'

// LIBS
import isEqual from 'react-fast-compare'
import {s} from 'react-native-size-matters/extend'

// Types
import {IProductCardProps} from '@model-types'

// Themes
import {Icons, Metrics} from '@themes'

// Styles
import {Image} from '@components/Image/Image.styles'
import {
  IconHeartWrapperStyled,
  ImageWrapperStyled,
  ProductCardWrapperStyled,
} from './ProductCard.styles'
import PStyled from '@components/Paragraph/P.styles'
import ViewStyled from '@components/View/View.styles'

// Constant
import {PARAGRAPH_TYPE} from '@constants'

const ProductCard = ({
  product,
  onPressProductCard,
  onPressLikeProduct,
}: IProductCardProps) => {
  const {url, like, name, type, price} = product || {}

  // handle action press product card with id
  const handlePressCardProduct = useCallback(() => {
    onPressProductCard(product.id)
  }, [product.id, onPressProductCard])

  // handle action press like product
  const handlePressLikeProduct = useCallback(() => {
    onPressLikeProduct(product)
  }, [product, onPressLikeProduct])

  return (
    <ProductCardWrapperStyled>
      <ImageWrapperStyled onPress={handlePressCardProduct}>
        <Image.Normal
          alignSelf="center"
          height={s(200)}
          width={s(160)}
          source={{
            uri: url,
          }}
          resizeMode="cover"
        />
        <IconHeartWrapperStyled onPress={handlePressLikeProduct}>
          <Image.Normal
            height={s(20)}
            width={s(20)}
            source={like ? Icons.heartLiked : Icons.heart}
            resizeMode="contain"
          />
        </IconHeartWrapperStyled>
      </ImageWrapperStyled>

      <PStyled.Base type={PARAGRAPH_TYPE.PRODUCT_NAME_AND_TYPE}>
        {name}
      </PStyled.Base>
      <ViewStyled.Custom mBottom={Metrics.margin.tiny}>
        <PStyled.Base type={PARAGRAPH_TYPE.PRODUCT_NAME_AND_TYPE}>
          {type}
        </PStyled.Base>
      </ViewStyled.Custom>
      <PStyled.Base type={PARAGRAPH_TYPE.PRODUCT_PRICE}>
        {`$ ${price}`}
      </PStyled.Base>
    </ProductCardWrapperStyled>
  )
}

export default memo(ProductCard, isEqual)
