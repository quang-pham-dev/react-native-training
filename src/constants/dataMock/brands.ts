export const brand = {
  id: '0',
  name: 'Nike',
  url: 'https://1000logos.net/wp-content/uploads/2021/11/Nike-Logo.png',
}

export const brands = [
  {
    id: '0',
    name: 'Nike',
    url: 'https://1000logos.net/wp-content/uploads/2021/11/Nike-Logo.png',
  },
  {
    id: '1',
    name: 'Adidas',
    url: 'https://www.pikpng.com/pngl/m/42-429866_adidas-logo-png-images-free-download-rh-pngimg.png',
  },
  {
    id: '2',
    name: 'Fila',
    url: 'https://toppng.com/uploads/preview/fila-logo-vector-download-free-11574216707jcmvfu58fy.png',
  },
  {
    id: '3',
    name: 'Puma',
    url: 'https://banner2.cleanpng.com/20180613/gsj/kisspng-tracksuit-puma-logo-brand-clothing-puma-logo-5b20dc6b232426.487874081528880235144.jpg',
  },
  {
    id: '4',
    name: 'Gucci',
    url: 'https://esquilo.io/png/thumb/XBwczcMncYeff65-Vector-Gucci-Logo-PNG-Free-Download.png',
  },
]
