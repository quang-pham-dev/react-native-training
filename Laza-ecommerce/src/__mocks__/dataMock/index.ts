import { user } from './user';
import { brand, brands } from './brands';
import { product, products } from './products';

export { brand, brands, product, products, user };
