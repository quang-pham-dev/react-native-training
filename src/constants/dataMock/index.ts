import {user} from './user'
import {brand, brands} from './brands'

export {brand, brands, user}
