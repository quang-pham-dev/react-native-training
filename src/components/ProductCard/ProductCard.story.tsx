import React from 'react'

// Storybook
import {action} from '@storybook/addon-actions'
import {StoryScreen, Story, UseCase} from '../../../storybook/views'
import {storiesOf} from '@storybook/react-native'

// Components
import ProductCard from './index'

// Constants
import {product} from '@constants'

declare let module: NodeModule

storiesOf('Product Card', module)
  .addDecorator(fn => <StoryScreen>{fn()}</StoryScreen>)
  .add('Default', () => (
    <Story>
      <UseCase text="default">
        <ProductCard
          product={product}
          onPressProductCard={action('clicked')}
          onPressLikeProduct={action('clicked')}
        />
      </UseCase>
    </Story>
  ))
