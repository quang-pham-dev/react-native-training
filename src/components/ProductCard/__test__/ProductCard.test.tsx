import React from 'react'

// LIBS
import {render} from '@testing-library/react-native'

// Components
import ProductCard from '@components/ProductCard'

// Mocks
import {product} from '@constants/dataMock'

describe('Product Card Component', () => {
  const props = {
    product,
    onPressProductCard: jest.fn(),
    onPressLikeProduct: jest.fn(),
  }

  const tree = render(<ProductCard {...props} />)
  test('Should render correctly', () => {
    const component = tree.toJSON()
    expect(component).toMatchSnapshot()
  })
})
