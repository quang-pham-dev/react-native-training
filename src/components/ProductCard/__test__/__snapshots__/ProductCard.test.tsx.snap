// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card Component Should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "marginBottom": 15,
      },
    ]
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "borderRadius": 15,
          "marginBottom": 5,
          "overflow": "hidden",
        },
      ]
    }
  >
    <Image
      alignSelf="center"
      height={200}
      resizeMode="cover"
      source={
        Object {
          "uri": "https://images.unsplash.com/photo-1618212179058-8fc73c26668d?ixlib=rb-1.2.1&raw_url=true&q=80&fm=jpg&crop=entropy&cs=tinysrgb&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=774",
        }
      }
      style={
        Array [
          Object {
            "alignSelf": "center",
            "height": 200,
            "resizeMode": "cover",
            "width": 160,
          },
        ]
      }
      width={160}
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "position": "absolute",
            "right": 15,
            "top": 14,
            "zIndex": 1,
          },
        ]
      }
    >
      <Image
        height={20}
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/assets/icons/heart-liked.webp",
          }
        }
        style={
          Array [
            Object {
              "height": 20,
              "resizeMode": "contain",
              "width": 20,
            },
          ]
        }
        width={20}
      />
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#1D1E20",
          "fontFamily": "Inter-Medium",
          "fontSize": 11,
          "lineHeight": 15.4,
          "textAlign": "left",
        },
      ]
    }
    type="productNameAndType"
  >
    Fila Sportswear Club Fleece
  </Text>
  <View
    mBottom={5}
    style={
      Array [
        Object {
          "marginBottom": 5,
          "width": "100%",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1D1E20",
            "fontFamily": "Inter-Medium",
            "fontSize": 11,
            "lineHeight": 15.4,
            "textAlign": "left",
          },
        ]
      }
      type="productNameAndType"
    >
      Cotton
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#1D1E20",
          "fontFamily": "Inter-SemiBold",
          "fontSize": 13,
          "lineHeight": 14.3,
          "textAlign": "left",
        },
      ]
    }
    type="productPrice"
  >
    $ 100
  </Text>
</View>
`;
